test/FxERC20.js

```diff
@@ -0,0 +1,209 @@
+            ).to.be.revertedWith("FxBaseChildTunnel: INVALID_SENDER");
+
+            // FxRoot as a sender is incorrect
+            await expect(
+                fxERC20ChildTunnel.connect(deployer).processMessageFromRoot(stateId, signers[1].address, "0x")
+            ).to.be.revertedWith("FxBaseChildTunnel: INVALID_SENDER_FROM_ROOT");
+        });
+
+        it("Deposit tokens", async function () {
+            // Approve tokens
+            await childToken.approve(fxERC20ChildTunnel.address, amount);
+
+            // Send tokens to L1
+            await fxERC20ChildTunnel.connect(deployer).deposit(amount);
+
+            // Tokens must be locked on the FxERC20ChildTunnel contract address
+            const balance = await childToken.balanceOf(fxERC20ChildTunnel.address);
+            expect(balance).to.equal(amount);
+
+            // On L1 fxERC20RootTunnel will be passed a proof validation data that the tx has happened on L2
+            await rootToken.mint(deployer.address, amount);
+        });
+
+        it("Deposit tokens to a different address", async function () {
+            const account = signers[1].address;
+
+            // Approve tokens
+            await childToken.approve(fxERC20ChildTunnel.address, amount);
+
+            // Send tokens to L1
+            await fxERC20ChildTunnel.connect(deployer).depositTo(account, amount);
+
+            // Tokens must be locked on the FxERC20ChildTunnel contract address
+            const balance = await childToken.balanceOf(fxERC20ChildTunnel.address);
+            expect(balance).to.equal(amount);
+
+            // On L1 fxERC20RootTunnel will be passed a proof validation data that the tx has happened on L2
+            await rootToken.mint(account, amount);
+        });
+
+        it("Withdraw tokens", async function () {
+            // Approve tokens
+            await childToken.connect(deployer).approve(fxERC20ChildTunnel.address, amount);
+
+            // Send tokens to L1
+            await fxERC20ChildTunnel.connect(deployer).deposit(amount);
+
+            // Tokens must be locked on the FxERC20ChildTunnel contract address
+            let balance = await childToken.balanceOf(fxERC20ChildTunnel.address);
+            expect(balance).to.equal(amount);
+
+            // On L1 fxERC20RootTunnel will be passed a proof validation data that the tx has happened on L2
+            await rootToken.mint(deployer.address, amount);
+
+            // Withdraw tokens
+            await rootToken.connect(deployer).approve(fxERC20RootTunnel.address, amount);
+
+            // Root token must be owned by the FxERC20RootTunnel contract
+            await rootToken.changeOwner(fxERC20RootTunnel.address);
+
+            // Burn tokens on L1 and send message to L2 to retrieve them there
+            await expect(
+                fxERC20RootTunnel.withdraw(amount)
+            ).to.be.reverted;
```

Comment:
why?

Reply:
The `FxERC20RootTunnel` ultimately leads to the `FxRoot` contract calling the cross-chain `StatSync`, which is impossible to recreate in a testing environment. I can try to write an `FxRootMock` contract for that purpose.

Reply:
FxRootMock is in place